src/plugins/kdb/ldap/libkdb_ldap/ldap_principal2.c

```diff
@@ -1375,64 +1492,198 @@ krb5_read_tkt_policy(krb5_context context, krb5_ldap_context *ldap_context,
+    unsigned int n_keys, i, ki, total_keys;
+    struct keyset *keysets = NULL;
+
+    *keysets_out = NULL;
+    *n_keysets_out = 0;
+    if (total_keys_out)
+        *total_keys_out = 0;
+
+    /* Precount the number of keys. */
+    for (n_keys = 0, i = 0; bvalues[i] != NULL; i++) {
+        if (bvalues[i]->bv_len > 0)
+            n_keys++;
+    }
 
-    if ((st=krb5_unparse_name(context, entries->princ, &user)) != 0)
+    keysets = k5calloc(n_keys, sizeof(struct keyset), &err);
+    if (keysets == NULL)
         goto cleanup;
 
-    for (i=0; bvalues[i] != NULL; ++i) {
-        krb5_int16 n_kd;
-        krb5_key_data *kd;
+    for (i = 0, ki = 0, total_keys = 0; bvalues[i] != NULL; i++) {
         krb5_data in;
 
         if (bvalues[i]->bv_len == 0)
             continue;
+
         in.length = bvalues[i]->bv_len;
         in.data = bvalues[i]->bv_val;
 
-        st = asn1_decode_sequence_of_keys (&in,
-                                           &kd,
-                                           &n_kd,
-                                           mkvno);
-
-        if (st != 0) {
-            const char *msg = error_message(st);
-            st = -1; /* Something more appropriate ? */
-            k5_setmsg(context, st,
-                      _("unable to decode stored principal key data (%s)"),
-                      msg);
-            goto cleanup;
-        }
-        noofkeys += n_kd;
-        tmp = key_data;
-        /* Allocate an extra key data to avoid allocating zero bytes. */
-        key_data = realloc(key_data, (noofkeys + 1) * sizeof (krb5_key_data));
-        if (key_data == NULL) {
-            key_data = tmp;
-            st = ENOMEM;
+        err = asn1_decode_sequence_of_keys(&in, &keysets[ki].key_data,
+                                           &keysets[ki].n_key_data, mkvno);
+        if (err)
             goto cleanup;
-        }
-        for (j = 0; j < n_kd; j++)
-            key_data[noofkeys - n_kd + j] = kd[j];
-        free (kd);
+
+        if (total_keys_out)
+            total_keys += keysets[ki].n_key_data;
+        ki++;
     }
 
-    entries->n_key_data = noofkeys;
+    if (total_keys_out)
+        *total_keys_out = total_keys;
+
+    *n_keysets_out = n_keys;
+    *keysets_out = keysets;
+    keysets = NULL;
+
+cleanup:
+    free(keysets);
+    return err;
+}
+
+krb5_error_code
+krb5_decode_krbsecretkey(krb5_context context, krb5_db_entry *entries,
+                         struct berval **bvalues, krb5_kvno *mkvno)
+{
+    char                        *user=NULL;
+    krb5_key_data               *key_data=NULL, *tmp;
+    krb5_error_code             err=0;
+
+    struct keyset *keys = NULL;
+    unsigned int i, n_key_entries, total_keys;
+
+    err = krb5_unparse_name(context, entries->princ, &user);
+    if (err)
+        goto cleanup;
+
+    err = decode_keys(bvalues, mkvno, &keys, &n_key_entries, &total_keys);
+    if (err != 0) {
+        const char *msg = error_message(err);
+        k5_setmsg(context, err,
```

Comment:
Since you're changing this code, make it use k5_prependmsg() (an internal alias for krb5_prepend_error_message()).